File: J9Ch/src/J_8_Data_Structure/Related_7/_606_Kth_Largest_Element_II_Medium.java
package J_8_Data_Structure.Related_7;
import java.util.*;
/** 606 Kth Largest Element II


 * Created by tianhuizhu on 6/28/17.
 */
public class _606_Kth_Largest_Element_II_Medium {

    class Solution {
        /**
         * @param nums an integer unsorted array
         * @param k an integer from 1 to n
         * @return the kth largest element
         */
        public int kthLargestElement2(int[] nums, int k) {
            // Write your code here
            PriorityQueue<Integer> q = new PriorityQueue<Integer>(k);
            for (int num : nums) {
                q.offer(num);
                if (q.size() > k) {
                    q.poll();
                }
            }
            return q.peek();
        }
    }
}

File: J9Ch/src/J_8_Data_Structure/Related_7/_606_Kth_Largest_Element_II_Medium.cpp
/** 606 Kth Largest Element II */


class Solution {
public:
    /**
     * @param nums an integer unsorted array
     * @param k an integer from 1 to n
     * @return the kth largest element
     */
    int kthLargestElement2(vector<int> nums, int k) {
        // Write your code here
        priority_queue<int, vector<int>, greater<int> > que;
        int n = nums.size();
        for(int i = 0; i < n; i ++) {
            if(que.size() == k) {
                if(nums[i] > que.top()) {
                    que.pop();
                    que.push(nums[i]);
                }
            }else {
                que.push(nums[i]);
            }
        }
        return que.top();
    }
};